test/integration/replicaset/replicaset_test.go

```diff
@@ -250,11 +494,9 @@ func TestAdoption(t *testing.T) {
 			t.Fatalf("Failed to create Pod: %v", err)
 		}
 
-		stopCh := make(chan struct{})
-		informers.Start(stopCh)
+		defer close(runControllerAndInformers(rm, informers))
```

Comment:
This works, but it's misleading at first. The reader will tend to see that the line starts with `defer` and assume it has no effects until the function is about to return.

To make it clear what's being deferred and that's happening now, it's better to keep these as separate lines:

```go
stopCh := runControllerAndInformers(rm, informers)
defer close(stopCh)
```

Separately, though, `runControllerAndInformers` might be too simple to be worth abstracting. Also, you've changed the behavior by moving `go rm.Run()` to happen before `waitToObservePods`. The order was important so that the shared informer has already observed the initial set of Pods (if any) before the controller starts.

Reply:
Yup, we should not change the order (although the test does pass). I will revert it back.
I still prefer abstracting the controller and informer setup codes though (like below as suggested). Will it be fine for you?
```
stopCh := runControllerAndInformers(rm, informers)
defer close(stopCh)
```

Reply:
If all the tests create their RSs and Pods before reaching this point, then it makes sense to keep `runControllerAndInformers` as a helper function. However, you will need to call `waitToObservePods` inside that helper, so it will need to know the expected number of initial pods.

Reply:
The test involving `waitToObservePods` does not have RS running before reaching the point mentioned, so the `runControllerAndInformers` will not be helpful. I will remove `runControllerAndInformers`.

Reply:
I didn't mean that the RS needs to be running before calling `runControllerAndInformers`. It's just that the Pod objects must have been created already. The helper seems fine as far as I can see.

Reply:
I see! The helper function triggers RS to run. I think the best way is to not use the helper function for the test involving `waitToObservePods`, but still using it for all other tests. Having `waitToObservePods` in the helper function requires knowing expected number of initial pods, which is unnecessary for all other tests.

Reply:
Based on offline sync, we should have  `waitToObservePods` call in the helper function. I verify everything works fine without additional delay.